File: src/main/java/org/meetkt/basket/domain/model/Basket.java
package org.meetkt.basket.domain.model;

import org.meetkt.catalogue.domain.model.Product;
import org.meetkt.catalogue.domain.model.ProductId;

import java.util.Arrays;
import java.util.Collections;
import java.util.Optional;

public class Basket {

    private final Items items;

    private Basket() {
        this.items = new Items(Collections.emptyList());
    }

    public static Basket empty() {
        return new Basket();
    }

    public void add(Product product) {
        this.items.add(new Item(product));
    }

    public void add(Product... products) {
        Arrays.stream(products).forEach(this::add);
    }

    public void deleteAllProductsMatching(ProductId productId) {
        this.items.removeIf(item -> item.contains(productId));
    }

    public Optional<Item> findBy(ProductId productId) {
        return items.findFirst(productId);
    }

    public int totalItems() {
        return this.items.size();
    }

    public Items allItems() {
        return this.items;
    }

    public int totalPrice() {
        return this.items.totalPrice();
    }
}

File: src/main/kotlin/in/meetkt/basket/domain/model/Basket.kt
package `in`.meetkt.basket.domain.model

import `in`.meetkt.catalogue.domain.model.Product
import `in`.meetkt.catalogue.domain.model.ProductId

class Basket {

    private val items: Items = Items(arrayListOf())

    companion object {
        fun empty(): Basket {
            return Basket()
        }
    }

    fun add(product: Product) {
        items += Item(product)
    }

    fun add(vararg products: Product) {
        products.forEach { items += Item(it) }
    }

    fun deleteAllProductsMatching(productId: ProductId) {
        items.removeIf { it.contains(productId) }
    }

    fun findBy(productId: ProductId): Item? = items.findFirst(productId)

    fun totalItems(): Int = items.size

    fun totalPrice(): Price = items.totalPrice()
}